### ISSUE TITLE
Incompatible with Gradle 7.6

### ISSUE THREAD

**Issue by Goooler** [asciidoctor/asciidoctor-gradle-plugin#655]
You can repro this issue on https://github.com/apache/groovy/pull/1832.

```
> Task :groovy-ant:asciidoctor
Dependency verification has been disabled for configuration detachedConfiguration2
Dependency verification has been disabled for configuration detachedConfiguration3
Error: Exception in thread "main" java.lang.NoClassDefFoundError: org/gradle/api/GradleException
	at org.gradle.internal.classpath.Instrumented.<clinit>(Instrumented.java:94)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec.$instrumentedCallSiteArray(AsciidoctorJavaExec.groovy)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec.$getCallSiteArray(AsciidoctorJavaExec.groovy)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec.registerExtensions(AsciidoctorJavaExec.groovy)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec.access$0(AsciidoctorJavaExec.groovy)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec$_run_closure2.doCall(AsciidoctorJavaExec.groovy:64)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec$_run_closure2.call(AsciidoctorJavaExec.groovy)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2359)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2344)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2385)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec.run(AsciidoctorJavaExec.groovy:62)
	at org.asciidoctor.gradle.remote.AsciidoctorJavaExec.main(AsciidoctorJavaExec.groovy:49)
Caused by: java.lang.ClassNotFoundException: org.gradle.api.GradleException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 12 more
```

**Comment by daniellansun** [comment]
ping @ysb33r  @aalmiray ;-)

**Comment by ysb33r** [comment]
Hey @daniellansun, please try to 4.0.0-alpha.1 release and tell me if it fixes the problem

**Comment by Goooler** [comment]
@ysb33r Seems not, see
https://github.com/Goooler/groovy/commit/bba5f5c0e15ef1451c600796b4bb33b47d9f1557
https://github.com/Goooler/groovy/actions/runs/3814781422/jobs/6489327740

**Comment by ysb33r** [comment]
@daniellansun It seems to be happening when extensions are deserialised. Since this is not happening with Gradle 7.5.1, but with Gradle  7.6, it might be possible that some Gradle dependencies are being leaked.

Do any of the Groovy asciidoc extensions rely on Gradle code?

**Comment by ysb33r** [comment]
@daniellansun That seems to be the problem. If I apply this diff to your repo, the build passes.

```diff
diff --git a/build-logic/src/main/groovy/org.apache.groovy-asciidoctor.gradle b/build-logic/src/main/groovy/org.apache.groovy-asciidoctor.gradle
index 7b1d4bb3c3..b827285d3a 100644
--- a/build-logic/src/main/groovy/org.apache.groovy-asciidoctor.gradle
+++ b/build-logic/src/main/groovy/org.apache.groovy-asciidoctor.gradle
@@ -158,6 +158,7 @@ def htmlOutputSanityCheck = { file, text, errors ->
 
 asciidoctor {
     def errors = new LinkedHashSet<String>()
+    inProcess IN_PROCESS
     doFirst {
         def specTestDir = file('src/spec/test')
         if (specTestDir.exists()) {
```

**Comment by daniellansun** [comment]
@ysb33r  Thanks a lot for your looking into the issue and giving the solution.
@Goooler Could you try again?

**Comment by Goooler** [comment]
Still failed, see
https://github.com/Goooler/groovy/commit/06d1e9b578654cc23a3cdd8f333cf41c295806b3
https://github.com/Goooler/groovy/actions/runs/3826304522/jobs/6509985057

**Comment by ysb33r** [comment]
Ah yes, that is the `asciidoctorPdf` task. Up to now we could not run that task `IN_PROCESS`, because it leaked an incompatible version of `snakeyaml` onto the classpath. I'll check to see if that is still the case.

Otherwise, this is going to be a much bigger fix.

**Comment by mrotteveel** [comment]
Another option is to add the Gradle API explicitly to the classpath. For example:

```
//...

configurations {
    asciidocExtensions
}

dependencies {
    asciidocExtensions gradleApi()
}

asciidoctor {
    configurations 'asciidocExtensions'
    // ...
}

asciidoctorPdf {
    configurations 'asciidocExtensions'
    // ...
}

// ...
```

**Comment by Goooler** [comment]
@mrotteveel Failed due to 
```
An exception occurred applying plugin request [id: 'org.apache.groovy-core']
> Failed to apply plugin 'org.apache.groovy-core'.
   > Failed to apply plugin 'org.apache.groovy-base'.
      > Failed to apply plugin 'org.apache.groovy-asciidoctor'.
         > org.gradle.internal.metaobject.AbstractDynamicObject$CustomMessageMissingMethodException: Could not find method asciidoctorPdf() for arguments [precompiled_OrgApacheGroovyAsciidoctor$_run_closure12@2b1d9078] on root project 'groovy' of type org.gradle.api.Project.
```
see
https://github.com/Goooler/groovy/commit/5b881c894fdf161f5e2e59ea372972d84fc25dcb
https://github.com/Goooler/groovy/actions/runs/3862782763/jobs/6584560388

**Comment by mrotteveel** [comment]
@Goooler That script doesn't seem to have `org.asciidoctor.jvm.pdf` as plugin, so remove the `asciidoctorPdf` definition.

**Comment by Goooler** [comment]
> @Goooler That script doesn't seem to have `org.asciidoctor.jvm.pdf` as plugin, so remove the `asciidoctorPdf` definition. My fix worked (with 3.3.2) in https://github.com/FirebirdSQL/jaybird-manual/blob/master/build.gradle

Oooops, I should fallback to 3.3.2 and patch your workaround, it works, thanks!.

**Comment by PacoVK** [comment]
@ysb33r is there yet a fix? Using 4.0.0-alpha.1 still breaks. I would contribute if possible but need at least a bit of context
